pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1456,18 +1456,42 @@ protected void handleFlow(CommandFlow flow) {
     @Override
     protected void handleRedeliverUnacknowledged(CommandRedeliverUnacknowledgedMessages redeliver) {
         checkArgument(state == State.Connected);
+        CompletableFuture<Consumer> consumerFuture = consumers.get(redeliver.getConsumerId());
+        final boolean hasRequestId = redeliver.hasRequestId();
+        final long requestId = hasRequestId ? redeliver.getRequestId() : 0;
+        final long consumerId = redeliver.getConsumerId();
+
         if (log.isDebugEnabled()) {
-            log.debug("[{}] Received Resend Command from consumer {} ", remoteAddress, redeliver.getConsumerId());
+            log.debug("[{}] redeliverUnacknowledged from consumer {} , requestId {}, consumerEpoch {}",
+                    remoteAddress, redeliver.getConsumerId(), requestId, redeliver.getConsumerEpoch());
         }
 
-        CompletableFuture<Consumer> consumerFuture = consumers.get(redeliver.getConsumerId());
-
         if (consumerFuture != null && consumerFuture.isDone() && !consumerFuture.isCompletedExceptionally()) {
             Consumer consumer = consumerFuture.getNow(null);
             if (redeliver.getMessageIdsCount() > 0 && Subscription.isIndividualAckMode(consumer.subType())) {
                 consumer.redeliverUnacknowledgedMessages(redeliver.getMessageIdsList());
             } else {
-                consumer.redeliverUnacknowledgedMessages();
+                consumer.redeliverUnacknowledgedMessages(redeliver.getConsumerEpoch()).whenComplete((v, e) -> {
+                    if (hasRequestId) {
+                        if (e != null) {
+                            log.error("redeliverUnacknowledgedMessages error! "
+                                    + "consumerId : {}, requestId: {}", consumerId, requestId, e);
+                            ctx.writeAndFlush(Commands.newError(requestId,
+                                    BrokerServiceException.getClientErrorCode(e), e.getMessage()));
+                        } else {
+                            ctx.writeAndFlush(Commands.newSuccess(requestId));
```

Comment:
Any reason for introducing the redelivery response here? If it is not directly related to the consumer epoch introduced in PIP 84, we can separate it into 2 parts.